File: v8-core/src/oracle/episode.rs
//! Oracle Benchmark Hardening & 3-Tier Episode Formulation (Issue #279, D-138).
//!
//! Owning Authority: V8 Constitution Rules 1, 4, 5, 6, 18; TARGET_ORACLE_SPEC §§1-11.
//!
//! Taxonomy & Hierarchy:
//!   - O0: Raw Horizon Indicator (diagnostic forward excursion per event)
//!   - O1: Episode Oracle (causal, non-overlapping directional swings; canonical denominator for recall)
//!   - O2: Feasible Portfolio Oracle (capital, concurrency, margin, and venue fees)

#![allow(dead_code)]

use serde::{Deserialize, Serialize};
use crate::hash::Canon;

/// 3-Tier Oracle Classification Level (D-138).
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub enum OracleTier {
    O0RawHorizon,
    O1Episode,
    O2FeasiblePortfolio,
}

/// Canonical parameters defining the Oracle measurement frame (D-138).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct OracleDefinition {
    pub definition_id: String,
    pub symbol: String,
    pub horizon_bars: usize,
    pub min_mfe_pct: f64,
    pub max_mae_pct: f64,
    pub min_rr_ratio: f64,
    pub roundtrip_friction_bps: f64,
    pub non_overlapping_policy: bool,
}

impl OracleDefinition {
    pub fn new(
        symbol: &str,
        horizon_bars: usize,
        min_mfe_pct: f64,
        max_mae_pct: f64,
        min_rr_ratio: f64,
        roundtrip_friction_bps: f64,
        non_overlapping_policy: bool,
    ) -> Self {
        let mut c = Canon::new();
        c.push_str("OracleDefinition-v1");
        c.push_str(symbol);
        c.push_u64(horizon_bars as u64);
        c.push_f64(min_mfe_pct);
        c.push_f64(max_mae_pct);
        c.push_f64(min_rr_ratio);
        c.push_f64(roundtrip_friction_bps);
        c.push_str(if non_overlapping_policy { "true" } else { "false" });
        let definition_id = c.finish_blake3_hex();

        Self {
            definition_id,
            symbol: symbol.to_string(),
            horizon_bars,
            min_mfe_pct,
            max_mae_pct,
            min_rr_ratio,
            roundtrip_friction_bps,
            non_overlapping_policy,
        }
    }
}

/// A structured non-overlapping Oracle Opportunity Episode (O1).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct OracleEpisode {
    pub episode_id: String,
    pub definition_id: String,
    pub symbol: String,
    pub direction: String, // "LONG" or "SHORT"
    pub entry_bar: usize,
    pub exit_bar: usize,
    pub duration_bars: usize,
    pub entry_price: f64,
    pub optimal_exit_price: f64,
    pub gross_mfe_pct: f64,
    pub gross_mae_pct: f64,
    pub gross_r: f64,
    pub habitat_type: String,
}

impl OracleEpisode {
    pub fn compute_id(&self) -> String {
        let mut c = Canon::new();
        c.push_str("OracleEpisode-v1");
        c.push_str(&self.definition_id);
        c.push_str(&self.symbol);
        c.push_str(&self.direction);
        c.push_u64(self.entry_bar as u64);
        c.push_u64(self.exit_bar as u64);
        c.push_f64(self.entry_price);
        c.push_f64(self.optimal_exit_price);
        c.finish_blake3_hex()
    }
}

/// Engine to extract canonical O1 episodes from historical price bars.
pub struct OracleEpisodeExtractor;

impl OracleEpisodeExtractor {
    pub fn extract_episodes(
        def: &OracleDefinition,
        highs: &[f64],
        lows: &[f64],
        closes: &[f64],
        volumes: &[f64],
        atrs: &[f64],
    ) -> Vec<OracleEpisode> {
        let n_bars = closes.len();
        let mut episodes = Vec::new();
        let mut i = 0;

        while i + def.horizon_bars < n_bars {
            let entry_price = closes[i];
            let atr = atrs.get(i).copied().unwrap_or(1.0).max(1e-6);

            let fwd_end = (i + def.horizon_bars + 1).min(n_bars);
            let fwd_highs = &highs[i + 1..fwd_end];
            let fwd_lows = &lows[i + 1..fwd_end];

            if fwd_highs.is_empty() || fwd_lows.is_empty() {
                i += 1;
                continue;
            }

            let mut max_h = fwd_highs[0];
            let mut max_h_rel_idx = 0;
            for (idx, &h) in fwd_highs.iter().enumerate() {
                if h > max_h {
                    max_h = h;
                    max_h_rel_idx = idx;
                }
            }

            let mut min_l = fwd_lows[0];
            let mut min_l_rel_idx = 0;
            for (idx, &l) in fwd_lows.iter().enumerate() {
                if l < min_l {
                    min_l = l;
                    min_l_rel_idx = idx;
                }
            }

            let max_h_bar = i + 1 + max_h_rel_idx;
            let min_l_bar = i + 1 + min_l_rel_idx;

            // Long metrics
            let long_mae_price = lows[i + 1..=max_h_bar].iter().copied().fold(f64::INFINITY, f64::min);
            let long_mfe_pct = (max_h - entry_price) / entry_price;
            let long_mae_pct = (entry_price - long_mae_price).max(0.0) / entry_price;

            // Short metrics
            let short_mae_price = highs[i + 1..=min_l_bar].iter().copied().fold(f64::NEG_INFINITY, f64::max);
            let short_mfe_pct = (entry_price - min_l) / entry_price;
            let short_mae_pct = (short_mae_price - entry_price).max(0.0) / entry_price;

            let mut admitted_ep = None;

            if long_mfe_pct >= def.min_mfe_pct
                && long_mae_pct <= def.max_mae_pct
                && (long_mfe_pct / long_mae_pct.max(0.001)) >= def.min_rr_ratio
            {
                let dur = max_h_bar.saturating_sub(i).max(1);
                let gross_r = (max_h - entry_price) / atr;
                let hab = if volumes.get(i).copied().unwrap_or(1.0) > 1.2 {
                    "TrendExpansionBreakout"
                } else {
                    "TrendPullbackContinuation"
                };

                let mut ep = OracleEpisode {
                    episode_id: String::new(),
                    definition_id: def.definition_id.clone(),
                    symbol: def.symbol.clone(),
                    direction: "LONG".to_string(),
                    entry_bar: i,
                    exit_bar: max_h_bar,
                    duration_bars: dur,
                    entry_price,
                    optimal_exit_price: max_h,
                    gross_mfe_pct: long_mfe_pct * 100.0,
                    gross_mae_pct: long_mae_pct * 100.0,
                    gross_r,
                    habitat_type: hab.to_string(),
                };
                ep.episode_id = ep.compute_id();
                admitted_ep = Some(ep);
            } else if short_mfe_pct >= def.min_mfe_pct
                && short_mae_pct <= def.max_mae_pct
                && (short_mfe_pct / short_mae_pct.max(0.001)) >= def.min_rr_ratio
            {
                let dur = min_l_bar.saturating_sub(i).max(1);
                let gross_r = (entry_price - min_l) / atr;
                let hab = if volumes.get(i).copied().unwrap_or(1.0) > 1.2 {
                    "TrendExpansionBreakout"
                } else {
                    "TrendPullbackContinuation"
                };

                let mut ep = OracleEpisode {
                    episode_id: String::new(),
                    definition_id: def.definition_id.clone(),
                    symbol: def.symbol.clone(),
                    direction: "SHORT".to_string(),
                    entry_bar: i,
                    exit_bar: min_l_bar,
                    duration_bars: dur,
                    entry_price,
                    optimal_exit_price: min_l,
                    gross_mfe_pct: short_mfe_pct * 100.0,
                    gross_mae_pct: short_mae_pct * 100.0,
                    gross_r,
                    habitat_type: hab.to_string(),
                };
                ep.episode_id = ep.compute_id();
                admitted_ep = Some(ep);
            }

            if let Some(ep) = admitted_ep {
                let next_step = if def.non_overlapping_policy { ep.exit_bar } else { i + 1 };
                episodes.push(ep);
                i = next_step;
            } else {
                i += 1;
            }
        }

        episodes
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_oracle_definition_deterministic_hashing() {
        let def1 = OracleDefinition::new("BTCUSDT", 24, 0.02, 0.01, 2.0, 10.0, true);
        let def2 = OracleDefinition::new("BTCUSDT", 24, 0.02, 0.01, 2.0, 10.0, true);
        let def3 = OracleDefinition::new("ETHUSDT", 24, 0.02, 0.01, 2.0, 10.0, true);

        assert_eq!(def1.definition_id, def2.definition_id);
        assert_ne!(def1.definition_id, def3.definition_id);
    }

    #[test]
    fn test_oracle_episode_extraction_and_non_overlap() {
        let def = OracleDefinition::new("BTCUSDT", 4, 0.02, 0.01, 2.0, 10.0, true);
        let highs = vec![100.0, 101.0, 103.0, 101.0, 100.0, 105.0, 102.0, 101.0];
        let lows = vec![99.0, 99.5, 99.5, 99.0, 98.0, 100.0, 100.0, 99.0];
        let closes = vec![100.0, 100.5, 102.5, 100.0, 99.0, 104.0, 101.0, 100.0];
        let volumes = vec![1.0; 8];
        let atrs = vec![1.0; 8];

        let eps = OracleEpisodeExtractor::extract_episodes(&def, &highs, &lows, &closes, &volumes, &atrs);
        assert!(!eps.is_empty());
        assert_eq!(eps[0].direction, "LONG");
        assert_eq!(eps[0].entry_bar, 0);
        assert_eq!(eps[0].exit_bar, 2);
        assert!(eps[0].compute_id().len() == 64);
    }
}

File: v8-next/src/v8_next/oracle/episode.py
"""O1 canonical episode extractor — port of v8-core/src/oracle/episode.rs (D-138).

Tiers (TARGET_ORACLE_SPEC §§1–11):
- O0 (raw horizon excursion) is a diagnostic only: ``o0_forward_excursion`` measures realized
  forward excursion per event for evaluators. Its output is labeled ``EVALUATOR_ONLY`` and
  must never be fed to decision features (R1).
- O1 (episode oracle): causal, non-overlapping directional swings — the canonical denominator
  for recall. Pure function over a polars frame (lib-first: polars owns the frame).
- O2 (feasible portfolio): capital/concurrency/margin/fees — not in this substrate.

DIVERGENCES from the Rust module: identity digests use hashlib (sha256) with the same
domain tags; reproducible inside this port, not bit-equal to the Rust blake3 Canon digests.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from enum import StrEnum
from typing import Sequence

import polars as pl

__all__ = [
    "EPISODE_FRAME_COLUMNS",
    "O0_EVALUATOR_ONLY",
    "OracleDefinition",
    "OracleEpisode",
    "OracleEpisodeExtractor",
    "OracleTier",
    "o0_forward_excursion",
]


class OracleTier(StrEnum):
    O0_RAW_HORIZON = "O0_RAW_HORIZON"
    O1_EPISODE = "O1_EPISODE"
    O2_FEASIBLE_PORTFOLIO = "O2_FEASIBLE_PORTFOLIO"


#: Columns the extractor reads from the input frame (all f64).
EPISODE_FRAME_COLUMNS = ("high", "low", "close", "volume", "atr")

#: Marker carried by every O0 diagnostic row: evaluator-only, never a decision feature.
O0_EVALUATOR_ONLY = "EVALUATOR_ONLY"


def _digest(domain: str, payload: object) -> str:
    blob = json.dumps(payload, sort_keys=True, separators=(",", ":"), allow_nan=False)
    return hashlib.sha256(f"{domain}|{blob}".encode()).hexdigest()


@dataclass(frozen=True)
class OracleDefinition:
    """Canonical parameters defining the O1 measurement frame (D-138)."""

    symbol: str
    horizon_bars: int
    min_mfe_pct: float
    max_mae_pct: float
    min_rr_ratio: float
    roundtrip_friction_bps: float
    non_overlapping_policy: bool
    definition_id: str = ""

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "definition_id",
            _digest(
                "OracleDefinition-v1",
                {
                    "symbol": self.symbol,
                    "horizon_bars": self.horizon_bars,
                    "min_mfe_pct": self.min_mfe_pct,
                    "max_mae_pct": self.max_mae_pct,
                    "min_rr_ratio": self.min_rr_ratio,
                    "roundtrip_friction_bps": self.roundtrip_friction_bps,
                    "non_overlapping_policy": self.non_overlapping_policy,
                },
            ),
        )


@dataclass(frozen=True)
class OracleEpisode:
    """One structured non-overlapping O1 opportunity episode."""

    episode_id: str
    definition_id: str
    symbol: str
    direction: str  # "LONG" or "SHORT"
    entry_bar: int
    exit_bar: int
    duration_bars: int
    entry_price: float
    optimal_exit_price: float
    gross_mfe_pct: float
    gross_mae_pct: float
    gross_r: float
    habitat_type: str

    def compute_id(self) -> str:
        return _digest(
            "OracleEpisode-v1",
            {
                "definition_id": self.definition_id,
                "symbol": self.symbol,
                "direction": self.direction,
                "entry_bar": self.entry_bar,
                "exit_bar": self.exit_bar,
                "entry_price": self.entry_price,
                "optimal_exit_price": self.optimal_exit_price,
            },
        )

    def as_dict(self) -> dict[str, object]:
        return {
            "episode_id": self.episode_id,
            "definition_id": self.definition_id,
            "symbol": self.symbol,
            "direction": self.direction,
            "entry_bar": self.entry_bar,
            "exit_bar": self.exit_bar,
            "duration_bars": self.duration_bars,
            "entry_price": self.entry_price,
            "optimal_exit_price": self.optimal_exit_price,
            "gross_mfe_pct": self.gross_mfe_pct,
            "gross_mae_pct": self.gross_mae_pct,
            "gross_r": self.gross_r,
            "habitat_type": self.habitat_type,
        }


def _columns(frame: pl.DataFrame) -> dict[str, list[float]]:
    missing = [c for c in EPISODE_FRAME_COLUMNS if c not in frame.columns]
    if missing:
        raise ValueError(f"episode frame is missing columns: {missing}")
    return {c: [float(v) for v in frame[c].to_list()] for c in EPISODE_FRAME_COLUMNS}


def o0_forward_excursion(
    frame: pl.DataFrame, horizon_bars: int
) -> list[dict[str, object]]:
    """O0 diagnostic: realized forward excursion per bar (evaluator-only).

    Returns ``EVALUATOR_ONLY`` rows; callers must never feed these to decision features.
    """
    cols = _columns(frame)
    closes, highs, lows = cols["close"], cols["high"], cols["low"]
    out: list[dict[str, object]] = []
    for i in range(len(closes)):
        fwd_end = min(i + horizon_bars + 1, len(closes))
        window_h = highs[i + 1 : fwd_end]
        window_l = lows[i + 1 : fwd_end]
        out.append(
            {
                "bar": i,
                "tier": OracleTier.O0_RAW_HORIZON.value,
                "usage": O0_EVALUATOR_ONLY,
                "forward_max_gain_pct": (max(window_h) - closes[i]) / closes[i] if window_h else None,
                "forward_max_loss_pct": (closes[i] - min(window_l)) / closes[i] if window_l else None,
            }
        )
    return out


class OracleEpisodeExtractor:
    """Canonical O1 episode extraction (pure function over a polars frame)."""

    @staticmethod
    def extract_episodes(definition: OracleDefinition, frame: pl.DataFrame) -> list[OracleEpisode]:
        cols = _columns(frame)
        highs, lows, closes = cols["high"], cols["low"], cols["close"]
        volumes, atrs = cols["volume"], cols["atr"]
        n_bars = len(closes)
        episodes: list[OracleEpisode] = []
        i = 0
        horizon = definition.horizon_bars
        while i + horizon < n_bars:
            entry_price = closes[i]
            atr = max(atrs[i] if i < len(atrs) else 1.0, 1e-6)
            fwd_end = min(i + horizon + 1, n_bars)
            fwd_highs = highs[i + 1 : fwd_end]
            fwd_lows = lows[i + 1 : fwd_end]
            if not fwd_highs or not fwd_lows:
                i += 1
                continue
            max_h = max(fwd_highs)
            max_h_bar = i + 1 + fwd_highs.index(max_h)
            min_l = min(fwd_lows)
            min_l_bar = i + 1 + fwd_lows.index(min_l)

            long_mae_price = min(lows[i + 1 : max_h_bar + 1])
            long_mfe_pct = (max_h - entry_price) / entry_price
            long_mae_pct = max(entry_price - long_mae_price, 0.0) / entry_price

            short_mae_price = max(highs[i + 1 : min_l_bar + 1])
            short_mfe_pct = (entry_price - min_l) / entry_price
            short_mae_pct = max(short_mae_price - entry_price, 0.0) / entry_price

            admitted: OracleEpisode | None = None
            volume_i = volumes[i] if i < len(volumes) else 1.0
            habitat = "TrendExpansionBreakout" if volume_i > 1.2 else "TrendPullbackContinuation"
            if (
                long_mfe_pct >= definition.min_mfe_pct
                and long_mae_pct <= definition.max_mae_pct
                and (long_mfe_pct / max(long_mae_pct, 0.001)) >= definition.min_rr_ratio
            ):
                duration = max(max_h_bar - i, 1)
                admitted = OracleEpisode(
                    episode_id="",
                    definition_id=definition.definition_id,
                    symbol=definition.symbol,
                    direction="LONG",
                    entry_bar=i,
                    exit_bar=max_h_bar,
                    duration_bars=duration,
                    entry_price=entry_price,
                    optimal_exit_price=max_h,
                    gross_mfe_pct=long_mfe_pct * 100.0,
                    gross_mae_pct=long_mae_pct * 100.0,
                    gross_r=(max_h - entry_price) / atr,
                    habitat_type=habitat,
                )
            elif (
                short_mfe_pct >= definition.min_mfe_pct
                and short_mae_pct <= definition.max_mae_pct
                and (short_mfe_pct / max(short_mae_pct, 0.001)) >= definition.min_rr_ratio
            ):
                duration = max(min_l_bar - i, 1)
                admitted = OracleEpisode(
                    episode_id="",
                    definition_id=definition.definition_id,
                    symbol=definition.symbol,
                    direction="SHORT",
                    entry_bar=i,
                    exit_bar=min_l_bar,
                    duration_bars=duration,
                    entry_price=entry_price,
                    optimal_exit_price=min_l,
                    gross_mfe_pct=short_mfe_pct * 100.0,
                    gross_mae_pct=short_mae_pct * 100.0,
                    gross_r=(entry_price - min_l) / atr,
                    habitat_type=habitat,
                )
            if admitted is not None:
                object.__setattr__(admitted, "episode_id", admitted.compute_id())
                episodes.append(admitted)
                i = admitted.exit_bar if definition.non_overlapping_policy else i + 1
            else:
                i += 1
        return episodes

    @staticmethod
    def extract_from_sequences(
        definition: OracleDefinition,
        highs: Sequence[float],
        lows: Sequence[float],
        closes: Sequence[float],
        volumes: Sequence[float],
        atrs: Sequence[float],
    ) -> list[OracleEpisode]:
        """Sequence entry point (tests/fixtures); the frame form above is canonical."""
        frame = pl.DataFrame(
            {
                "high": [float(v) for v in highs],
                "low": [float(v) for v in lows],
                "close": [float(v) for v in closes],
                "volume": [float(v) for v in volumes],
                "atr": [float(v) for v in atrs],
            }
        )
        return OracleEpisodeExtractor.extract_episodes(definition, frame)
